modules/apps/portal-background-task/portal-background-task-service/src/main/java/com/liferay/portal/background/task/internal/BackgroundTaskMessagingConfigurator.java

```diff
@@ -0,0 +1,141 @@
+
+/**
+ * @author Vendel Toreki
+ */
+@Component(
+	immediate = true, service = BackgroundTaskMessagingConfigurator.class
+)
+public class BackgroundTaskMessagingConfigurator {
+
+	@Activate
+	protected void activate(BundleContext bundleContext) {
+		_bundleContext = bundleContext;
+
+		Destination backgroundTaskDestination = _registerDestination(
+			bundleContext, DestinationConfiguration.DESTINATION_TYPE_PARALLEL,
+			DestinationNames.BACKGROUND_TASK, 5, 10);
+
+		backgroundTaskDestination.register(
+			new BackgroundTaskMessageListener(
+				_backgroundTaskExecutorRegistry, _backgroundTaskLocalService,
+				_backgroundTaskStatusRegistry,
+				_backgroundTaskThreadLocalManager, _lockManager, _messageBus));
+
+		Destination backgroundTaskStatusDestination = _registerDestination(
+			bundleContext, DestinationConfiguration.DESTINATION_TYPE_SERIAL,
+			DestinationNames.BACKGROUND_TASK_STATUS, 1, 1);
+
+		backgroundTaskStatusDestination.register(
+			new BackgroundTaskQueuingMessageListener(
+				_backgroundTaskLocalService, _lockManager));
+		backgroundTaskStatusDestination.register(
+			new RemoveOnCompletionBackgroundTaskStatusMessageListener(
+				_backgroundTaskLocalService));
+	}
+
+	@Deactivate
+	protected void deactivate() {
+		for (ServiceRegistration<Destination> serviceRegistration :
+				_serviceRegistrations) {
+
+			Destination destination = _bundleContext.getService(
+				serviceRegistration.getReference());
+
+			serviceRegistration.unregister();
+
+			destination.destroy();
+		}
+
+		_bundleContext = null;
+	}
+
+	private Destination _registerDestination(
+		BundleContext bundleContext, String destinationType,
+		String destinationName, int workersCoreSize, int workersMaxSize) {
+
+		DestinationConfiguration destinationConfiguration =
+			new DestinationConfiguration(destinationType, destinationName);
+
+		destinationConfiguration.setWorkersCoreSize(workersCoreSize);
+		destinationConfiguration.setWorkersMaxSize(workersMaxSize);
+
+		Destination destination = _destinationFactory.createDestination(
+			destinationConfiguration);
+
+		_serviceRegistrations.add(
+			bundleContext.registerService(
+				Destination.class, destination,
+				HashMapDictionaryBuilder.<String, Object>put(
+					"destination.name", destination.getName()
+				).build()));
+
+		return destination;
+	}
+
+	@Reference
+	private BackgroundTaskExecutorRegistry _backgroundTaskExecutorRegistry;
+
+	@Reference
+	private BackgroundTaskLocalService _backgroundTaskLocalService;
+
+	@Reference
+	private BackgroundTaskStatusRegistry _backgroundTaskStatusRegistry;
+
+	@Reference
+	private BackgroundTaskThreadLocalManager _backgroundTaskThreadLocalManager;
+
+	private volatile BundleContext _bundleContext;
+
+	@Reference
+	private DestinationFactory _destinationFactory;
+
+	@Reference
+	private LockManager _lockManager;
+
+	@Reference
+	private MessageBus _messageBus;
+
+	private final Set<ServiceRegistration<Destination>> _serviceRegistrations =
```

Comment:
@rafaprax this supposed to keep track of ServiceRegistration for unregistering, there can not be any duplication on ServiceRegistrations, this should be a plain List for better performance